### PR TITLE
8320379: C2: Sort spilling/unspilling sequence for better ld/st merging into ldp/stp on AArch64

### PR DESCRIPTION
**Pull request by fg1417**
Macro-assembler on aarch64 can merge adjacent loads or stores into ldp/stp.[[1]](https://github.com/openjdk/jdk/blob/a95062b39a431b4937ab6e9e73de4d2b8ea1ac49/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L2079)

For example, it can merge:
```
str     w20, [sp, #16]
str     w10, [sp, #20]
```
into
```
stp     w20, w10, [sp, #16]
```

But C2 may generate a sequence like:
```
str     x21, [sp, #8]
str     w20, [sp, #16]
str     x19, [sp, #24] <---
str     w10, [sp, #20] <--- Before sorting
str     x11, [sp, #40]
str     w13, [sp, #48]
str     x16, [sp, #56]
```
We can't do any merging for non-adjacent loads or stores.

The patch is to sort the spilling or unspilling sequence in the order of offset during instruction scheduling and bundling phase. After that, we can get a new sequence:
```
str     x21, [sp, #8]
str     w20, [sp, #16]
str     w10, [sp, #20] <---
str     x19, [sp, #24] <--- After sorting
str     x11, [sp, #40]
str     w13, [sp, #48]
str     x16, [sp, #56]
```

Then macro-assembler can do ld/st merging:
```
str     x21, [sp, #8]
stp     w20, w10, [sp, #16] <--- Merged
str     x19, [sp, #24]
str     x11, [sp, #40]
str     w13, [sp, #48]
str     x16, [sp, #56]
```

To justify the patch, we run `HelloWorld.java`
```
public class HelloWorld {
    public static void main(String [] args) {
        System.out.println("Hello World!");
    }
}
```
with `java -Xcomp -XX:-TieredCompilation HelloWorld`.

Before the patch, macro-assembler can do ld/st merging for 3688 times. After the patch, the number of ld/st merging increases to 3871 times, by ~5 %.

Tested tier1~3 on x86 and AArch64.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320379](https://bugs.openjdk.org/browse/JDK-8320379): C2: Sort spilling/unspilling sequence for better ld/st merging into ldp/stp on AArch64 (**Enhancement** - P4)

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16754/head:pull/16754` \
`$ git checkout pull/16754`

Update a local copy of the PR: \
`$ git checkout pull/16754` \
`$ git pull https://git.openjdk.org/jdk.git pull/16754/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16754`

View PR using the GUI difftool: \
`$ git pr show -t 16754`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16754.diff">https://git.openjdk.org/jdk/pull/16754.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16754#issuecomment-1820365898)

### CODE REVIEW

File: src/hotspot/share/opto/output.cpp
Comment (theRealAph) [inline src/hotspot/share/opto/output.cpp:2297]:
```suggestion
    return _regalloc->reg2offset(first_dst_lo) - _regalloc->reg2offset(second_dst_lo);
```

File: src/hotspot/share/opto/output.cpp
Comment (theRealAph) [inline src/hotspot/share/opto/output.cpp:175]:
```suggestion
  // Return an integer less than, equal to, or greater than zero if the stack offset of the
  // first argument is respectively less than, equal to, or greater than the second.
 
```

File: src/hotspot/share/opto/output.cpp
Comment (theRealAph) [inline src/hotspot/share/opto/output.cpp:2324]:
```suggestion
               compare_two_spill_nodes(n, _available[i]) > 0) {
```

File: src/hotspot/share/opto/output.cpp
Comment (theRealAph) [inline src/hotspot/share/opto/output.cpp:2316]:
```suggestion
  // them in the order of stack offset. Some ports (e.g. aarch64) may also
```

File: src/hotspot/share/opto/output.cpp
Comment (theRealAph) [inline src/hotspot/share/opto/output.cpp:2291]:
```suggestion
    return _regalloc->reg2offset(first_src_lo) - _regalloc->reg2offset(second_src_lo);
```

File: src/hotspot/share/opto/output.cpp
Comment (theRealAph) [inline src/hotspot/share/opto/output.cpp:2300]:
```suggestion
  return 0; // Not comparable.
```

**Comment by theRealAph** [comment]
This is a good idea, although the real-world gains are small. I'd wonder if this was worth doing for non-AArch64 ports, although even on others sorting the accesses into order might help.

**Comment by fg1417** [comment]
Hi @theRealAph , thanks a lot for your review! All comments have been resolved in the new commit.

> This is a good idea, although the real-world gains are small. I'd wonder if this was worth doing for non-AArch64 ports, although even on others sorting the accesses into order might help.

Yeah, that's also bothering me. I'm not sure if it benefits other ports. Do you think if we need convert the change to aarch64-only? Thanks.

**Comment by theRealAph** [comment]
> Hi @theRealAph , thanks a lot for your review! All comments have been resolved in the new commit.
> 
> > This is a good idea, although the real-world gains are small. I'd wonder if this was worth doing for non-AArch64 ports, although even on others sorting the accesses into order might help.
> 
> Yeah, that's also bothering me. I'm not sure if it benefits other ports. Do you think if we need convert the change to aarch64-only? Thanks.

#ifdefs are probably wrose, so I'd leave it as it is. We need another reviewer.

**Comment by fg1417** [comment]
@theRealAph thanks for your review!

May I have another review? Maybe @vnkozlov? Thanks.

**Review by vnkozlov — COMMENTED** [review]
Looks reasonable. It may help other platform's pre-fetchers because you are ordering memory access.
I will run testing before approval.

**Review by vnkozlov — APPROVED** [review]
My tier1-4,xcomp,stress testing passed.

**Comment by fg1417** [comment]
Thanks for all your reviewing and test work! @vnkozlov @theRealAph 

I'll integrate it if there is no other comment.

**Pull request closed without merging**
